### PR TITLE
[New] `no-danger`: add `customComponentNames` option

### LINKED ISSUE #3434: feat: no-danger find props on all components (not just native)
Unless I am mistaken, it seems today that the rule `react/no-danger` has specific logic to only find usages on native elements. See the `isDOMComponent` check here:

https://github.com/jsx-eslint/eslint-plugin-react/blob/8c9ec8a3929403ac34a22c11cb9bd2575ba766bd/lib/rules/no-danger.js#L63

We think it could be useful to support non-native components as well, especially in cases like `styled-components` wrappers, or components that want to offer this prop to be forwarded internally to a native element.

This could be a new default behavior, or an opt-in behavior via a new rule option.

### PR DESCRIPTION
**Pull request by akulsr0**
Closes #3434

### CODE REVIEW

File: docs/rules/no-danger.md
Comment (ljharb) [inline docs/rules/no-danger.md:44]: i wonder if it'd be simpler to have just an array that defaults to `[]`, but can be set to `['*']` if you want to check everything?

File: lib/rules/no-danger.js
Comment (ljharb) [inline lib/rules/no-danger.js:82]: i didn't intend for `*` to be a special value here; i think we should use existing globbing logic to allow `Foo*` or `*Foo` or `*` etc.

File: lib/rules/no-danger.js
Comment (ljharb) [inline lib/rules/no-danger.js:66]:
```suggestion
          minItems: 0,
```
so that a config can reset the value to the default

File: lib/rules/no-danger.js
Comment (ljharb) [inline lib/rules/no-danger.js:83]:
```suggestion
        const enableCheckingCustomComponent = customComponentNames.some((name) => minimatch(functionName, name));
```
.some already handles empty lists :-)

File: tests/lib/rules/no-danger.js
Comment (ljharb) [inline tests/lib/rules/no-danger.js:53]:
```suggestion
      options: [{ customComponentNames: ['*MUI'] }],
```

Comment (ljharb) [reply tests/lib/rules/no-danger.js:53]: aha, ok thanks. can you add the same exact test case with the matching pattern, as well?

Comment (akulsr0) [reply tests/lib/rules/no-danger.js:53]: I have added that test case [here](https://github.com/jsx-eslint/eslint-plugin-react/pull/3748/files#diff-4b8ada5e2441fe2af247dda8f99e5c6d99df1b6cbe622f21c8dc6d7cef18ceadR90-R103)

Comment (ljharb) [reply tests/lib/rules/no-danger.js:53]: the test is good, but i'm suggesting having *identical* cases where the only difference is the glob pattern

Comment (akulsr0) [reply tests/lib/rules/no-danger.js:53]: Added same testcase with opposite pattern in invalid cases.

**Review by ljharb — APPROVED** [review]

**Comment by akulsr0** [comment]
Let me know if anything is pending here else can we proceed towards merging it?

**Comment by ljharb** [comment]
it's good to land; i want to cut a patch release before landing new semver-minor things, is all :-)

**Review by ljharb — APPROVED** [review]
LGTM, but i'm going to hold off landing new semver-minors for a bit.

**Comment by akulsr0** [comment]
Bumping this up

**Pull request merged**
